database/manager.go

```diff
@@ -0,0 +1,181 @@
+		go keyManager.run()
+		keyManagers[key] = keyManager
+	}
+	return &DatabaseManager{
+		logger:      logger,
+		db:          db,
+		interval:    interval,
+		keyManagers: keyManagers,
+	}
+}
+
+// Run writes all committed keys to the database on the configured timer.
+// This function should only be called once.
+func (dm *DatabaseManager) Run() {
+	for range time.Tick(dm.interval) {
+		for id, km := range dm.keyManagers {
+			// Ensure we're not writing the default value
+			if km.maxCommittedHeight > 0 {
+				storedHeight, err := getLatestProcessedBlockHeight(dm.db, id)
+				if err != nil && !IsKeyNotFoundError(err) {
+					dm.logger.Error("Failed to get latest processed block height", zap.Error(err))
+					continue
+				}
+				if storedHeight >= km.maxCommittedHeight {
+					continue
+				}
+				dm.logger.Debug(
+					"writing height",
+					zap.Uint64("height", km.maxCommittedHeight),
+					zap.String("relayerID", id.ID.String()),
+				)
+				err = dm.db.Put(id.ID, LatestProcessedBlockKey, []byte(strconv.FormatUint(km.maxCommittedHeight, 10)))
+				if err != nil {
+					dm.logger.Error("Failed to write latest processed block height", zap.Error(err))
+					continue
+				}
+			}
+		}
+	}
+}
+
+func (dm *DatabaseManager) Get(id RelayerID, key DataKey) ([]byte, error) {
+	return dm.db.Get(id.ID, key)
+}
+
+func (dm *DatabaseManager) PrepareHeight(id RelayerID, height uint64, totalMessages uint64) error {
+	km, ok := dm.keyManagers[id]
+	if !ok {
+		dm.logger.Error("Key manager not found", zap.String("relayerID", id.ID.String()))
+		return fmt.Errorf("key manager not found")
+	}
+	km.prepareHeight(height, totalMessages)
+	return nil
+}
+
+func (dm *DatabaseManager) Finished(id RelayerID, height uint64) {
+	km, ok := dm.keyManagers[id]
+	if !ok {
+		// This is not an error, as this will occur if the database is not eligible for writing
+		dm.logger.Debug("Key manager not found", zap.String("relayerID", id.ID.String()))
+		return
+	}
+	km.finished <- height
+}
+
+//
+// keyManager commits keys to be written to the database in a thread safe manner.
+//
+
+type keyManager struct {
+	logger                   logging.Logger
+	queuedHeightsLock        *sync.Mutex
+	queuedHeightsAndMessages map[uint64]*messageCounter
+	maxHeightLock            *sync.Mutex
+	maxCommittedHeight       uint64
+	finished                 chan uint64
+}
+
+func newKeyManager(logger logging.Logger) *keyManager {
+	return &keyManager{
+		logger:                   logger,
+		queuedHeightsLock:        &sync.Mutex{},
+		queuedHeightsAndMessages: make(map[uint64]*messageCounter),
+		maxHeightLock:            &sync.Mutex{},
+		finished:                 make(chan uint64),
+	}
+}
+
+// Run listens for finished signals from application relayers, and commits the
+// height once all messages have been processed.
+// This function should only be called once.
+func (km *keyManager) run() {
+	for height := range km.finished {
+		counter, ok := km.queuedHeightsAndMessages[height]
+		if !ok {
```

Comment:
should this ever happen? From what I can tell, the height should be committed if it were finished. If so, can you log an error?